Sources/SwiftOperatorPrecedence/PrecedenceGraph.swift

```diff
@@ -0,0 +1,149 @@
+
+  /// Add a new precedence group
+  ///
+  /// - throws: If there is already a precedence group with the given name,
+  ///   throws PrecedenceGraphError.groupAlreadyExists.
+  mutating func add(
+    _ group: PrecedenceGroup,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows {
+    if let existing = precedenceGroups[group.name] {
+      try errorHandler(
+        OperatorPrecedenceError.groupAlreadyExists(
+          existing: existing, new: group))
+    } else {
+      precedenceGroups[group.name] = group
+    }
+  }
+
+  /// Look for the precedence group with the given name, or return nil if
+  /// no such group is known.
+  func lookupGroup(_ groupName: PrecedenceGroupName) -> PrecedenceGroup? {
+    return precedenceGroups[groupName]
+  }
+
+  /// Search the precedence-group relationships, starting at the given
+  /// (fromGroup, fromSyntax) and following precedence groups in the
+  /// specified direction.
+  private func searchRelationships(
+    initialGroupName: PrecedenceGroupName, initialSyntax: SyntaxProtocol?,
+    targetGroupName: PrecedenceGroupName,
+    direction: PrecedenceRelation.Kind,
+    errorHandler: OperatorPrecedenceErrorHandler
+  ) rethrows -> Precedence? {
+    // Keep track of all of the groups we have seen during our exploration of
+    // the graph. This detects cycles and prevents extraneous work.
+    var groupsSeen: Set<PrecedenceGroupName> = []
+
+    var stack: [(PrecedenceGroupName, SyntaxProtocol?)] =
+      [(initialGroupName, initialSyntax)]
+    while let (currentGroupName, currentGroupSyntax) = stack.popLast() {
+      guard let currentGroup = lookupGroup(currentGroupName) else {
+        try errorHandler(
+          .missingGroup(currentGroupName, referencedFrom: currentGroupSyntax))
+        continue
+      }
+
+      for relation in currentGroup.relations {
+        if relation.kind == direction {
+          // If we hit our initial group, we're done.
+          let otherGroupName = relation.groupName
+          if otherGroupName == targetGroupName {
+            switch direction {
+            case .lowerThan:
+              return .lowerThan
+
+            case .higherThan:
+              return .higherThan
+            }
+          }
+
+          if groupsSeen.insert(otherGroupName).inserted {
+            stack.append((otherGroupName, relation.syntax))
+          }
+        }
+      }
+    }
+
+    return nil
+  }
+
+  /// Determine the precedence relationship between two precedence groups.
+  ///
+  /// Follow the precedence relationships among the precedence groups to
+  /// determine the precedence of the start group relative to the end group.
+  ///
+  /// - Returns: Precedence.lowerThan if startGroupName has lower precedence
+  /// than endGroupName, Precedence.higherThan if startGroupName has higher
+  /// precedence than endGroup name, and Precedence.unrelated otherwise.
+  func precedence(
```

Comment:
https://github.com/apple/swift-syntax/pull/619#discussion_r952214981 still applies here. I think a doc comment describing how `startSyntax` and `endSyntax` relate to `startGroupName` and `endGroupName` would be useful here.